File: crates/fairy-intelligence/src/store/extraction.rs
use super::*;

impl IntelligenceStore {
    pub fn pending_extraction_turn_count(
        &self,
        conversation_id: ConversationId,
    ) -> Result<u64, FairyError> {
        let count: i64 = self
            .lock()?
            .query_row(
                "SELECT COUNT(*) FROM conversation_turns
                 WHERE conversation_id = ?1 AND status = 'completed'
                   AND extraction_state = 'pending'",
                [conversation_id.to_string()],
                |row| row.get(0),
            )
            .map_err(|_| storage_error("无法统计待抽取 turn"))?;
        u64::try_from(count).map_err(|_| storage_error("待抽取 turn 数量已损坏"))
    }

    pub fn claim_extraction_batch(
        &self,
        conversation_id: ConversationId,
        limit: usize,
    ) -> Result<Option<ExtractionBatchInput>, FairyError> {
        if !(1..=12).contains(&limit) {
            return Err(invalid_record("抽取批次 limit 必须位于 1..=12"));
        }
        let now = now_unix_ms()?;
        let batch_id = ExtractionBatchId::new();
        let mut connection = self.lock()?;
        let transaction = connection
            .transaction()
            .map_err(|_| storage_error("无法开始抽取批次领取事务"))?;
        let character_id: String = transaction
            .query_row(
                "SELECT character_id FROM conversations WHERE id = ?1",
                [conversation_id.to_string()],
                |row| row.get(0),
            )
            .optional()
            .map_err(|_| storage_error("无法读取抽取批次角色"))?
            .ok_or_else(|| invalid_record("抽取 conversation 不存在"))?;
        let character_id = CharacterId::from_str(&character_id)
            .map_err(|_| storage_error("抽取批次 character id 已损坏"))?;
        let mut statement = transaction
            .prepare(
                "SELECT t.id, t.sequence, u.content, a.content
                 FROM conversation_turns t
                 JOIN conversation_messages u ON u.turn_id = t.id AND u.role = 'user'
                 JOIN conversation_messages a ON a.turn_id = t.id AND a.role = 'assistant'
                 WHERE t.conversation_id = ?1 AND t.status = 'completed'
                   AND t.extraction_state = 'pending'
                 ORDER BY t.sequence ASC
                 LIMIT ?2",
            )
            .map_err(|_| storage_error("无法准备待抽取 turn 查询"))?;
        let rows = statement
            .query_map(params![conversation_id.to_string(), limit as i64], |row| {
                Ok((
                    row.get::<_, String>(0)?,
                    row.get::<_, i64>(1)?,
                    row.get::<_, String>(2)?,
                    row.get::<_, String>(3)?,
                ))
            })
            .map_err(|_| storage_error("无法读取待抽取 turn"))?;
        let mut claimed = Vec::new();
        for row in rows {
            let (turn_id, sequence, user_message, assistant_message) =
                row.map_err(|_| storage_error("待抽取 turn 已损坏"))?;
            claimed.push((
                TurnId::from_str(&turn_id).map_err(|_| storage_error("待抽取 turn id 已损坏"))?,
                i64_to_u64(sequence, "待抽取 turn sequence 已损坏")?,
                user_message,
                assistant_message,
            ));
        }
        drop(statement);
        let Some(first) = claimed.first() else {
            return Ok(None);
        };
        let last = claimed.last().expect("claimed has first and last");
        transaction
            .execute(
                "INSERT INTO extraction_batches(
                    id, conversation_id, character_id, status,
                    first_turn_sequence, last_turn_sequence, created_at_ms, updated_at_ms
                 ) VALUES (?1, ?2, ?3, 'running', ?4, ?5, ?6, ?6)",
                params![
                    batch_id.to_string(),
                    conversation_id.to_string(),
                    character_id.to_string(),
                    u64_to_i64(first.1)?,
                    u64_to_i64(last.1)?,
                    now,
                ],
            )
            .map_err(|_| invalid_record("当前 conversation 已有 running 抽取批次"))?;
        for (turn_id, sequence, _, _) in &claimed {
            let changed = transaction
                .execute(
                    "UPDATE conversation_turns SET extraction_state = 'claimed', updated_at_ms = ?3
                     WHERE id = ?1 AND conversation_id = ?2 AND extraction_state = 'pending'",
                    params![turn_id.to_string(), conversation_id.to_string(), now],
                )
                .map_err(|_| storage_error("无法领取待抽取 turn"))?;
            if changed != 1 {
                return Err(invalid_record("待抽取 turn 已被其他批次领取"));
            }
            transaction
                .execute(
                    "INSERT INTO extraction_batch_turns(batch_id, turn_id, turn_sequence)
                     VALUES (?1, ?2, ?3)",
                    params![
                        batch_id.to_string(),
                        turn_id.to_string(),
                        u64_to_i64(*sequence)?
                    ],
                )
                .map_err(|_| storage_error("无法记录抽取批次 turn"))?;
        }
        let query = claimed
            .iter()
            .flat_map(|(_, _, user, assistant)| [user.as_str(), assistant.as_str()])
            .collect::<Vec<_>>()
            .join(" ");
        let mut remaining_chars = MAX_RETRIEVED_CONTEXT_CHARS;
        let existing_memories = match build_fts_query(&query)? {
            Some(fts_query) => {
                retrieve_personal(&transaction, character_id, &fts_query, &mut remaining_chars)?
            }
            None => Vec::new(),
        };
        transaction
            .commit()
            .map_err(|_| storage_error("无法提交抽取批次领取事务"))?;
        Ok(Some(ExtractionBatchInput {
            batch_id,
            conversation_id,
            character_id,
            turns: claimed
                .into_iter()
                .map(
                    |(turn_id, _, user_message, assistant_message)| ExtractionTurn {
                        turn_id,
                        user_message,
                        assistant_message,
                    },
                )
                .collect(),
            existing_memories,
        }))
    }

    pub fn fail_extraction_batch(
        &self,
        batch_id: ExtractionBatchId,
        error: &FairyError,
    ) -> Result<(), FairyError> {
        let changed = self
            .lock()?
            .execute(
                "UPDATE extraction_batches SET status = 'failed', error_code = ?2,
                    error_message = ?3, error_retryable = ?4, updated_at_ms = ?5
                 WHERE id = ?1 AND status = 'running'",
                params![
                    batch_id.to_string(),
                    error.code.as_str(),
                    error.message,
                    i64::from(error.retryable),
                    now_unix_ms()?
                ],
            )
            .map_err(|_| storage_error("无法记录抽取批次失败"))?;
        if changed == 1 {
            Ok(())
        } else {
            Err(invalid_record("抽取批次不存在或不是 running"))
        }
    }

    pub fn retry_failed_extraction_batch(
        &self,
        batch_id: ExtractionBatchId,
    ) -> Result<ConversationId, FairyError> {
        let now = now_unix_ms()?;
        let mut connection = self.lock()?;
        let transaction = connection
            .transaction()
            .map_err(|_| storage_error("无法开始失败批次重试事务"))?;
        let conversation_id: String = transaction
            .query_row(
                "SELECT conversation_id FROM extraction_batches
                 WHERE id = ?1 AND status = 'failed'",
                [batch_id.to_string()],
                |row| row.get(0),
            )
            .optional()
            .map_err(|_| storage_error("无法读取失败抽取批次"))?
            .ok_or_else(|| invalid_record("抽取批次不存在或不是 failed"))?;
        transaction
            .execute(
                "UPDATE conversation_turns SET extraction_state = 'pending', updated_at_ms = ?2
                 WHERE id IN (
                    SELECT turn_id FROM extraction_batch_turns WHERE batch_id = ?1
                 ) AND extraction_state = 'claimed'",
                params![batch_id.to_string(), now],
            )
            .map_err(|_| storage_error("无法释放失败批次 turn"))?;
        transaction
            .execute(
                "UPDATE extraction_batches SET status = 'cancelled', updated_at_ms = ?2
                 WHERE id = ?1 AND status = 'failed'",
                params![batch_id.to_string(), now],
            )
            .map_err(|_| storage_error("无法关闭失败抽取批次"))?;
        transaction
            .commit()
            .map_err(|_| storage_error("无法提交失败批次重试事务"))?;
        ConversationId::from_str(&conversation_id)
            .map_err(|_| storage_error("失败批次 conversation id 已损坏"))
    }
}

File: fairy/memory/extraction.go
package memory

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"unicode"

	"github.com/jackc/pgx/v5"
)

type ExtractionTurnRow struct {
	ID        string
	Sequence  int64
	User      string
	Assistant string
}

func BuildExtractionRetrievalProjection(turns []ExtractionTurn) []string {
	fragments := make([]string, 0, len(turns)*2)
	remaining := MaxFTSQueryChars
	for _, turn := range turns {
		for _, field := range []string{turn.UserMessage, turn.AssistantMessage} {
			for _, token := range strings.FieldsFunc(field, unicode.IsSpace) {
				token = strings.TrimSpace(token)
				runes := []rune(token)
				if len(runes) > extractionProjectionFragmentRunes {
					runes = runes[:extractionProjectionFragmentRunes]
				}
				if len(runes) > remaining {
					runes = runes[:remaining]
				}
				fragment := string(runes)
				if fragment == "" {
					continue
				}
				usable, err := BuildFTSQuery(fragment)
				if err != nil || usable == "" {
					continue
				}
				fragments = append(fragments, fragment)
				remaining -= len(runes)
				break
			}
			if remaining == 0 {
				return fragments
			}
		}
	}
	return fragments
}

func LockConversationCharacter(ctx context.Context, tx pgx.Tx, conversationID string) (string, error) {
	var characterID string
	if err := tx.QueryRow(ctx, "SELECT character_id FROM conversations WHERE id = $1 FOR UPDATE", conversationID).Scan(&characterID); errors.Is(err, pgx.ErrNoRows) {
		return "", errors.New("conversation does not exist")
	} else if err != nil {
		return "", fmt.Errorf("reading conversation character: %w", err)
	}
	return characterID, nil
}

func BuildExtractionBatchInput(
	ctx context.Context,
	tx pgx.Tx,
	batchID, conversationID, characterID string,
	claimed []ExtractionTurnRow,
	normalizeQuery func(string) (string, error),
) (*ExtractionBatchInput, error) {
	turns := make([]ExtractionTurn, 0, len(claimed))
	for _, item := range claimed {
		turns = append(turns, ExtractionTurn{TurnID: item.ID, UserMessage: item.User, AssistantMessage: item.Assistant})
	}
	remaining := maxRetrievedContextChars
	projection := BuildExtractionRetrievalProjection(turns)
	existing, err := RetrievePersonalExtractionProjection(ctx, tx, characterID, projection, &remaining, normalizeQuery)
	if err != nil {
		return nil, err
	}
	return &ExtractionBatchInput{
		BatchID:          batchID,
		ConversationID:   conversationID,
		CharacterID:      characterID,
		Turns:            turns,
		ExistingMemories: existing,
	}, nil
}

func SucceedExtractionBatch(ctx context.Context, tx pgx.Tx, batchID, workerID string, now int64) error {
	return succeedExtractionClaim(ctx, tx, batchID, workerID, now)
}

func LockRunningExtractionBatch(ctx context.Context, tx pgx.Tx, batchID, workerID string) (conversationID, characterID string, err error) {
	err = tx.QueryRow(ctx, `
SELECT turn.conversation_id, conversation.character_id
FROM conversation_turns AS turn
JOIN conversations AS conversation ON conversation.id = turn.conversation_id
WHERE turn.extraction_claim_id = $1
  AND turn.extraction_state = 'claimed'
  AND turn.extraction_lease_owner = $2
ORDER BY turn.sequence
LIMIT 1
FOR UPDATE OF turn`, batchID, workerID).Scan(&conversationID, &characterID)
	if errors.Is(err, pgx.ErrNoRows) {
		return "", "", errors.New("extraction batch does not exist or is not running")
	}
	if err != nil {
		return "", "", fmt.Errorf("reading running extraction batch: %w", err)
	}
	return conversationID, characterID, nil
}

func LoadExtractionBatchTurnIDs(ctx context.Context, tx pgx.Tx, batchID string) (map[string]struct{}, error) {
	rows, err := tx.Query(ctx, `
SELECT id
FROM conversation_turns
WHERE extraction_claim_id = $1 AND extraction_state = 'claimed'`, batchID)
	if err != nil {
		return nil, fmt.Errorf("loading extraction batch turns: %w", err)
	}
	defer rows.Close()
	allowedTurnIDs := make(map[string]struct{})
	for rows.Next() {
		var turnID string
		if err := rows.Scan(&turnID); err != nil {
			return nil, fmt.Errorf("scanning extraction batch turn: %w", err)
		}
		allowedTurnIDs[turnID] = struct{}{}
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterating extraction batch turns: %w", err)
	}
	return allowedTurnIDs, nil
}

func FindDuplicateMemory(ctx context.Context, db Querier, kind string, scope MemoryScope, content string) (string, error) {
	scopeKind, characterID, _ := MemoryScopeColumns(scope)
	var rows pgx.Rows
	var err error
	if characterID == nil {
		rows, err = db.Query(ctx, `
SELECT id, content FROM personal_memories
WHERE kind = $1 AND scope_kind = $2 AND character_id IS NULL
  AND status = 'active' AND review_status = 'ready'
ORDER BY updated_at_ms DESC, id ASC`, kind, scopeKind)
	} else {
		rows, err = db.Query(ctx, `
SELECT id, content FROM personal_memories
WHERE kind = $1 AND scope_kind = $2 AND character_id = $3
  AND status = 'active' AND review_status = 'ready'
ORDER BY updated_at_ms DESC, id ASC`, kind, scopeKind, *characterID)
	}
	if err != nil {
		return "", fmt.Errorf("querying duplicate memories: %w", err)
	}
	defer rows.Close()
	normalized := NormalizeMemoryContent(content)
	for rows.Next() {
		var id, existing string
		if err := rows.Scan(&id, &existing); err != nil {
			return "", fmt.Errorf("scanning duplicate memory: %w", err)
		}
		if NormalizeMemoryContent(existing) == normalized {
			return id, nil
		}
	}
	if err := rows.Err(); err != nil {
		return "", fmt.Errorf("iterating duplicate memories: %w", err)
	}
	return "", nil
}

func LinkPersonalMemoryEvidence(ctx context.Context, tx pgx.Tx, memoryID, turnID string, now int64) error {
	rows, err := tx.Query(ctx, `SELECT evidence_id FROM conversation_turn_evidence WHERE turn_id = $1 ORDER BY evidence_id`, turnID)
	if err != nil {
		return fmt.Errorf("loading memory source evidence: %w", err)
	}
	evidenceIDs := make([]string, 0, 8)
	for rows.Next() {
		var evidenceID string
		if err := rows.Scan(&evidenceID); err != nil {
			rows.Close()
			return fmt.Errorf("scanning memory source evidence: %w", err)
		}
		evidenceIDs = append(evidenceIDs, evidenceID)
	}
	if err := rows.Err(); err != nil {
		rows.Close()
		return fmt.Errorf("iterating memory source evidence: %w", err)
	}
	rows.Close()
	encoded, err := json.Marshal(evidenceIDs)
	if err != nil {
		return fmt.Errorf("serializing personal memory evidence: %w", err)
	}
	changed, err := tx.Exec(ctx, `
UPDATE personal_memories
SET evidence_ids_json = $2::jsonb, updated_at_ms = GREATEST(updated_at_ms, $3)
WHERE id = $1`, memoryID, encoded, now)
	if err != nil {
		return fmt.Errorf("storing personal memory evidence: %w", err)
	}
	if changed.RowsAffected() != 1 {
		return errors.New("personal memory does not exist")
	}
	return nil
}

func InsertMemoryContextCoverage(
	ctx context.Context,
	tx pgx.Tx,
	conversationID, turnID, memoryID, resultStatus string,
	now int64,
) error {
	if resultStatus != "applied" && resultStatus != "no_change" {
		return errors.New("memory context coverage result status is invalid")
	}
	changed, err := tx.Exec(ctx, `
INSERT INTO memory_context_coverages(
  conversation_id, turn_id, memory_id, result_status, created_at_ms
)
VALUES ($1, $2, $3, $4, $5)
ON CONFLICT (conversation_id, turn_id, memory_id)
DO UPDATE SET result_status = excluded.result_status`,
		conversationID, turnID, memoryID, resultStatus, now)
	if err != nil {
		return fmt.Errorf("inserting memory context coverage: %w", err)
	}
	if changed.RowsAffected() != 1 {
		return errors.New("memory context coverage was not committed")
	}
	return nil
}

func RequireActiveMemoryScope(ctx context.Context, tx pgx.Tx, memoryID, kind string, scope MemoryScope) error {
	current, err := PersonalMemoryByID(ctx, tx, memoryID, true)
	if err != nil {
		return err
	}
	if current.Status != "active" || current.Kind != kind || current.Scope.Type != scope.Type || current.Scope.CharacterID != scope.CharacterID {
		return errors.New("supersede target memory status, kind, or scope does not match")
	}
	return nil
}

func ListExtractionBatches(ctx context.Context, db Querier, characterID, status string) ([]ExtractionBatchRecord, error) {
	rows, err := db.Query(ctx, `
SELECT
  COALESCE(turn.extraction_claim_id, turn.id),
  turn.conversation_id,
  conversation.character_id,
  CASE WHEN turn.extraction_state = 'claimed' THEN 'running' ELSE turn.extraction_state END,
  min(turn.sequence),
  max(turn.sequence),
  min(COALESCE(turn.extraction_error_code, '')),
  min(COALESCE(turn.extraction_error_message, '')),
  min(turn.created_at_ms),
  max(turn.updated_at_ms)
FROM conversation_turns AS turn
JOIN conversations AS conversation ON conversation.id = turn.conversation_id
WHERE conversation.character_id = $1
  AND (CASE WHEN turn.extraction_state = 'claimed' THEN 'running' ELSE turn.extraction_state END) = $2
GROUP BY COALESCE(turn.extraction_claim_id, turn.id), turn.conversation_id, conversation.character_id,
         CASE WHEN turn.extraction_state = 'claimed' THEN 'running' ELSE turn.extraction_state END
ORDER BY max(turn.updated_at_ms) DESC, COALESCE(turn.extraction_claim_id, turn.id)
LIMIT 20`, characterID, status)
	if err != nil {
		return nil, fmt.Errorf("querying extraction batches: %w", err)
	}
	defer rows.Close()
	records := make([]ExtractionBatchRecord, 0)
	for rows.Next() {
		var record ExtractionBatchRecord
		var first, last int64
		var code, message string
		if err := rows.Scan(&record.ID, &record.ConversationID, &record.CharacterID, &record.Status, &first, &last, &code, &message, &record.CreatedAtUnixMS, &record.UpdatedAtUnixMS); err != nil {
			return nil, fmt.Errorf("scanning extraction batch: %w", err)
		}
		record.FirstTurnSequence = uint64(first)
		record.LastTurnSequence = uint64(last)
		if code != "" && message != "" {
			record.Error = &WireError{Code: code, Message: message, Retryable: true}
		}
		records = append(records, record)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterating extraction batches: %w", err)
	}
	return records, nil
}

func RetryFailedExtractionBatch(ctx context.Context, tx pgx.Tx, id string, now int64) (string, error) {
	var conversationID string
	if err := tx.QueryRow(ctx, `
SELECT conversation_id
FROM conversation_turns
WHERE id = $1 AND extraction_state = 'failed'
FOR UPDATE`, id).Scan(&conversationID); errors.Is(err, pgx.ErrNoRows) {
		return "", errors.New("extraction batch is not retryable")
	} else if err != nil {
		return "", fmt.Errorf("reading failed extraction batch: %w", err)
	}
	changed, err := tx.Exec(ctx, `
UPDATE conversation_turns
SET extraction_state = 'pending',
    extraction_attempt_count = 0,
    extraction_next_attempt_at_ms = 0,
    extraction_error_code = NULL,
    extraction_error_message = NULL,
    updated_at_ms = $2
WHERE id = $1 AND extraction_state = 'failed'`, id, now)
	if err != nil {
		return "", fmt.Errorf("cancelling failed extraction batch: %w", err)
	}
	if changed.RowsAffected() != 1 || conversationID == "" {
		return "", errors.New("extraction batch is not retryable")
	}
	return conversationID, nil
}
